### ISSUE TITLE
Unified Run and Deploy Button

### ISSUE THREAD

**Issue by dominiklohmann** [tenzir/public-roadmap#67]
Currently in the Explorer, users must decide whether they want to _Run_ or _Deploy_ a pipeline. However, at most one of the two actions makes sense at any given time, given that _Run_ requires a void to events and _Deploy_ a void to void pipeline.

We want to unify the two buttons into a single _Run_ button, and move the logic for deciding which action to take into Tenzir itself.

```[tasklist]
### Definition of Done
- [x] Design UX of the button for a pipeline that can be run or deployed, respectively.
- [x] Implement changes in Tenzir.
- [ ] Adapt to new API in the Explorer.
```

**Comment by tobim** [comment]
I'm a bit concerned about the loss of discoverability of a feature - namely creating new managed pipelines - that is core to the user value.

I do agree that the deploy button in it's current form is sub-optimal, but at least it is a visible interactive element that somehow signals to the user "here is the place to connect thing together".

Without going too much into the solution space, I think as a user I would appreciate a top-level entry point that handholds me through the process of deploying a managed pipeline. For example some kind of wizard that lets me choose source, sink and transformation. instead of creating a managed pipeline directly the "deploy" button would send the the user to the same wizard but with the source and transformation filled in from the editor contents.

**Comment by mavam** [comment]
What you describe sounds like an entirely new roadmap item, e.g., "Pipeline Wizard". I see that complementary to this issue. What you describe very much resembles the "visual operator picker" that I sketched out in an early Excalidraw. Basically compose your pipeline with a few clicks, instead of writing one in TQL.

Solution space idea to get closer to the desired effect within the scope of this story: in order to make the deploy scenario a bit more distinctive, we could change the color of the button once a sink is present.

**Comment by mavam** [comment]
> Such functionality was met with opposition before (mainly from @dit7ya), because whether a sink is present can only be decided after we have parsed the pipeline, which requires a call to the backend. Thus, we would have to constantly submit the pipeline to the backend and there would be a delay until the button changes colors.

If the limitation are only user-defined operators, then we have something that's 90% there and good enough until we have better TQL parsing. I'm not sure what the other reasons were, if any.

**Comment by tobim** [comment]
> Solution space idea to get closer to the desired effect within the scope of this story: in order to make the deploy scenario a bit more distinctive, we could change the color of the button once a sink is present.

I don't think this is enough. My point is that the current proposal reduces the discoverability of the deploy function and that makes the app worse.

**Comment by mavam** [comment]
We (@annadehtiarova, @jachris, @tobim, and me) discussed how to proceed here.

After agreeing that we want to make the UX context-dependent, i.e., a function of the presence of a sink, we explored several pathways:

1. Use a single button that changes its shape based on the presence of
   1. a sink
   2. an error
2. Make the sink more explicit, e.g., by introducing a dedicated line below Editor where Explorer would be one possible sink
3. Use a toggle-like button that flips between the Run and Deploy state. Basically (1) but with words.
4. Offer a toggle to switch between a code (= Editor) and graphical pipeline representation akin to [Elixir's Explorer](https://www.youtube.com/watch?v=U6nuPjyAUPw).

We ultimately agreed on (1) as viable path forward and reserved (4) for later exploration. A primary concern was that we did not want to split the pipeline into pieces, but rather treat it as unit in itself—like a piece of code.

A related topic that we need to address in the design is how we want to render errors, i.e., when it's neither possible to run nor to deploy a pipeline. Should we use the results pane? Or just underline the faulty parts in the pipeline with an error on hover?

Finally, @jachris mentioned an evolutionary idea of live-updating the results as the user types, completely ditching the run/play button. We would like to explore this idea in a separate roadmap item.
